runtime/src/accounts.rs

```diff
@@ -320,36 +332,96 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, account);
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &Arc<RwLock<HashMap<Pubkey, CreditOnlyLock>>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
     ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+        let credit_only_locks_read = credit_only_locks.read().unwrap();
+        let mut lock_counts: HashMap<&Pubkey, MutexGuard<u64>> = HashMap::new();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k) {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            } else if let Some(credit_only_lock) = credit_only_locks_read.get(&k) {
+                let lock_count = credit_only_lock.lock_count.lock().unwrap();
+                if *lock_count > 0 {
+                    error_counters.account_in_use += 1;
+                    debug!("Account in use: {:?}", k);
+                    return Err(TransactionError::AccountInUse);
+                } else {
+                    lock_counts.insert(k, lock_count);
+                }
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
+            } else if let Some(credit_only_lock) = credit_only_locks_read.get(&k) {
+                let lock_count = credit_only_lock.lock_count.lock().unwrap();
+                lock_counts.insert(k, lock_count);
+            }
+        }
+
+        let mut credit_only_writes: Vec<&Pubkey> = vec![];
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+            lock_counts.remove(k);
+        }
+        for k in credit_only_keys {
```

Comment:
Out of curiosity, if somebody included . the same key multiple times in a transaction such that it was both a credit-only account and a credit-debit account, where in the pipeline would that fail?

Reply:
The key can only be one thing or the other in a properly constructed Transaction (see MessageHeader)